crates/matrix-sdk/src/client/builder.rs

```diff
@@ -371,63 +402,88 @@ impl ClientBuilder {
-                {
-                    sliding_sync_proxy =
-                        self.sliding_sync_proxy.as_ref().map(|url| Url::parse(url)).transpose()?;
+        let (homeserver, well_known) = match homeserver_cfg {
+            HomeserverConfig::Url(url) => (url, None),
+            HomeserverConfig::ServerName { server: server_name, protocol } => {
+                let well_known = discover_homeserver(server_name, protocol, &http_client).await?;
+                (well_known.homeserver.base_url.clone(), Some(well_known))
+            }
+            HomeserverConfig::ServerNameOrUrl(server_name_or_url) => {
+                // Store the result to return at the end. If this doesn't get modified, then the
+                // supplied name is neither a server name, nor a valid URL.
+                let mut result: Result<
+                    (String, Option<discover_homeserver::Response>),
+                    ClientBuildError,
+                > = Err(ClientBuildError::InvalidServerName);
+
+                // Attempt discovery as a server name first.
+                let sanitize_result = sanitize_server_name(&server_name_or_url);
+                if let Ok(server_name) = sanitize_result.as_ref() {
+                    let protocol = if server_name_or_url.starts_with("http://") {
+                        UrlScheme::Http
+                    } else {
+                        UrlScheme::Https
+                    };
+
+                    result = match discover_homeserver(server_name.clone(), protocol, &http_client)
+                        .await
+                    {
+                        Ok(well_known) => {
+                            Ok((well_known.homeserver.base_url.clone(), Some(well_known)))
+                        }
+                        Err(e) => Err(e),
+                    }
                 }
 
-                #[cfg(feature = "experimental-oidc")]
-                {
-                    allow_insecure_oidc = url.starts_with("http://");
+                // When discovery fails, or the input isn't a valid server name, fallback to
+                // trying a homeserver URL if supplied.
+                if let Ok(homeserver_url) = Url::parse(&server_name_or_url) {
+                    // Make sure the URL is definitely for a homeserver.
+                    if http_client
+                        .send(
+                            get_supported_versions::Request::new(),
+                            Some(RequestConfig::short_retry()),
+                            homeserver_url.to_string(),
+                            None,
+                            &[MatrixVersion::V1_0],
+                            Default::default(),
+                        )
+                        .await
+                        .is_ok()
+                    {
+                        result = Ok((homeserver_url.to_string(), None));
+                    } else {
+                        // If it isn't, we'll throw the discovery error instead.
```

Comment:
Can you instead use a match and log the error at a `debug!` level, please?

Reply:
Ah nice, had a match initially but clippy didn't like it with just the comment. 👍